### ISSUE TITLE
btrfs check does not catch truncated filesystems

### ISSUE THREAD

**Issue by crass** [kdave/btrfs-progs#522]
When shrinking the underlying block device that a btrfs volume resides on, the user may accidentally shrink the underlying block device to a size smaller than the btrfs volume potentially causing filesystem corruption. Doing a `btrfs check` does not catch this, however the filesystem will fail to mount. I believe `btrfs check` should fail if the size of the block device is smaller than the btrfs filesystem size.

Here's a series of commands illustrating this:
```
uname -a; btrfs version
truncate -s 2G btrfs.img
mkfs.btrfs -f btrfs.img
mount btrfs.img /mnt
dd if=/dev/zero of=/mnt/1G.file count=2M
btrfs dev usage /mnt/
umount /mnt
truncate -s -1G btrfs.img
btrfs check -p --check-data-csum btrfs.img
echo "btrfs check returned: $?"
mount btrfs.img /mnt
dmesg | tail
umount /mnt
```

And here is the output I get from running these commands:
```
Linux datastore 5.19.0-15-generic #15-Ubuntu SMP PREEMPT_DYNAMIC Tue Aug 2 07:35:59 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
btrfs-progs v5.16.2
btrfs-progs v5.16.2
See http://btrfs.wiki.kernel.org for more information.

NOTE: several default settings have changed in version 5.15, please make sure
      this does not affect your deployments:
      - DUP for metadata (-m dup)
      - enabled no-holes (-O no-holes)
      - enabled free-space-tree (-R free-space-tree)

Label:              (null)
UUID:               34601ace-ef5d-4c63-9023-5c39cda0ef0c
Node size:          16384
Sector size:        4096
Filesystem size:    2.00GiB  
Block group profiles:
  Data:             single            8.00MiB
  Metadata:         DUP             102.38MiB
  System:           DUP               8.00MiB
SSD detected:       no
Zoned device:       no
Incompat features:  extref, skinny-metadata, no-holes
Runtime features:   free-space-tree
Checksum:           crc32c
Number of devices:  1
Devices:
   ID        SIZE  PATH
    1     2.00GiB  btrfs.img

2097152+0 records in
2097152+0 records out
1073741824 bytes (1.1 GB, 1.0 GiB) copied, 9.88967 s, 109 MB/s
/dev/loop0, ID: 1
   Device size:             2.00GiB
   Device slack:              0.00B
   Data,single:             1.02GiB
   Metadata,DUP:          204.75MiB
   System,DUP:             16.00MiB
   Unallocated:           779.25MiB

Opening filesystem to check...
Checking filesystem on btrfs.img
UUID: 34601ace-ef5d-4c63-9023-5c39cda0ef0c
[1/7] checking root items                      (0:00:00 elapsed, 104 items check[1/7] checking root items                      (0:00:00 elapsed, 104 items checked)
[2/7] checking extents                         (0:00:00 elapsed, 5 items checked[2/7] checking extents                         (0:00:00 elapsed, 94 items checked)
[3/7] checking free space tree                 (0:00:00 elapsed, 1 items checked[3/7] checking free space tree                 (0:00:00 elapsed, 11 items checked)
[4/7] checking fs roots                        (0:00:00 elapsed, 1 items checked[4/7] checking fs roots                        (0:00:00 elapsed, 2 items checked)
[5/7] checking csums against data              (0:00:00 elapsed, 1 items checked[5/7] checking csums against data              (0:00:01 elapsed, 49 items checke[5/7] checking csums against data              (0:00:01 elapsed, 72 items checked)
[6/7] checking root refs                       (0:00:00 elapsed, 3 items checked[6/7] checking root refs                       (0:00:00 elapsed, 3 items checked)
[7/7] checking quota groups skipped (not enabled on this FS)
found 1075134464 bytes used, no error found
total csum bytes: 1048576
total tree bytes: 1392640
total fs tree bytes: 32768
total extent tree bytes: 16384
btree space waste bytes: 314948
file data blocks allocated: 1073741824
 referenced 1073741824
btrfs check returned: 0
mount: /mnt: wrong fs type, bad option, bad superblock on /dev/loop0, missing codepage or helper program, or other error.
[480158.084167] BTRFS info (device loop0): has skinny extents
[480158.087413] BTRFS info (device loop0): enabling ssd optimizations
[480158.087694] BTRFS info (device loop0): checking UUID tree
[480171.405753] loop0: detected capacity change from 0 to 2097152
[480171.411211] BTRFS info (device loop0): flagging fs with big metadata feature
[480171.411222] BTRFS info (device loop0): using free space tree
[480171.411226] BTRFS info (device loop0): has skinny extents
[480171.412801] BTRFS error (device loop0): device total_bytes should be at most 1073741824 but found 2147483648
[480171.424470] BTRFS error (device loop0): failed to read chunk tree: -22
[480171.484953] BTRFS error (device loop0): open_ctree failed
umount: /mnt: not mounted.
```

**Comment by luontola** [comment]
I think this is the same issue as on the Ubuntu issue tracker: https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1931790

One workaround mentioned there is to downgrade to kernel 5.10 or older, so that you can mount the volume, and then running btrfs filesystem resize max. Another workaround, which I haven't tried: remove the problematic device, so that you can mount the filesystem, and then add the problematic disk back to the mounted filesystem and run btrfs filesystem resize max. (The error message doesn't say which device has the problem, so you need to check the partition sizes manually.)

Here is my experience report:
https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1931790/comments/20

TLDR:
I had the "device total_bytes should be at most X but found Y" error which prevented me from mounting the btrfs filesystem. It was apparently caused by a HDD's usable disk space/partition size shrinking by about 1MB (perhaps the HDD detected bad sectors). "btrfs check" found no errors. I had to downgrade to kernel 5.10 (with ubuntu-mainline-kernel.sh) to be able to mount the filesystem. After it was mounted, I was able to fix the filesystem by running "btrfs filesystem resize max". Then I could again boot to kernel 5.15.

**Comment by adam900710** [comment]
Nope, downgrading is not the correct way to go.

And unless you're just upgrading from old kernels, such situation should be a big warning.
(If you really believe it's some weird HDD behavior, I'd say the disk is not reliable at all)

I'll add such detection for `btrfs check`.

**Comment by adam900710** [comment]
@luontola  Nope, downgrading is not the correct way to go.

Furthermore, if your HDD is doing weird things, I strongly doubt if it would even handle FLUSH correctly, thus it's a ticking bomb, and a sudden powerloss/kernel panic can easily destroy your fs.
If you can not trust the hardware (well, a lot of times they are not trustworthy though), you should not use it.

And it has been some time after we implemented the strict size check, if you still hit it without doing huge kernel update, it is definitely something wrong.

For the `btrfs check` support, it's already implemented by commit 0a348d3c81e1 ("btrfs-progs: check: verify the underlying block device size is valid"), and commit e47c34821f3a ("btrfs-progs: rescue: allow fix-device-size to shrink device item") introduces the way to fix such case.
